riscv/base_cfg.cc

```diff
@@ -0,0 +1,79 @@
+// See LICENSE for license details.
+
+#include "base_cfg.h"
+#include "config.h"
+
+#include <cassert>
+
+void base_cfg_t::set_initrd_bounds(reg_t start, reg_t end)
+{
+  assert(start <= end);
+  initrd_start = start;
+  initrd_end = end;
+}
+
+void base_cfg_t::set_bootargs(const char *str)
+{
+  bootargs = str;
+}
+
+void base_cfg_t::set_nprocs(size_t n)
+{
+  assert(n > 0);
+  nprocs = n;
+}
+
+void base_cfg_t::set_isa(const char *str)
+{
+  assert(str);
+  isa = str;
+}
+
+void base_cfg_t::set_priv(const char *str)
+{
+  assert(str);
+  priv = str;
+}
+
+void base_cfg_t::set_mems(const std::vector<std::pair<reg_t, mem_t*>> &m,
+                          bool from_arg)
+{
+  mems.clear();
+  mems.reserve(m.size());
+  for (const auto &pr : m) {
+    mems.push_back(std::make_pair(pr.first, pr.second->size()));
+  }
+  explicit_mems = from_arg;
+}
+
+std::pair<reg_t, reg_t> base_cfg_t::get_initrd_bounds() const
+{
+  return std::make_pair(initrd_start, initrd_end);
+}
+
+std::pair<size_t, bool> base_cfg_t::get_nprocs() const
+{
+  size_t num = (nprocs == 0) ? 1 : nprocs;
+  bool was_explicit = (nprocs != 0);
```

Comment:
This smells funny, having default-illegal values to detect explicit vs implicit. It seems like the defaults should all be specified as a group in one place, and probably not in this class, instead of spread throughout this code, and then you could use `std::optional` to keep track of which ones were set explicitly?  (Except that `std::optional` is C++17 and we're [currently compiling](https://github.com/riscv-software-src/riscv-isa-sim/blob/59ec157568d2a52feeec568ac042362db1c5ddbc/Makefile.in#L97) Spike with `-std=c++11` :frowning_face:)